pkg/gcv/validator.go

```diff
@@ -155,8 +197,51 @@ func (v *Validator) AddData(request *validator.AddDataRequest) error {
 	return nil
 }
 
+// Review evaluates each asset in the review request in parallel and returns any
+// violations found.
 func (v *Validator) Review(ctx context.Context, request *validator.ReviewRequest) (*validator.ReviewResponse, error) {
-	return nil, errors.Errorf("Not implemented")
+	var requestDone sync.WaitGroup
+	var responseMutex sync.Mutex
+	response := &validator.ReviewResponse{}
+	var errs multierror.Errors
+
+	requestDone.Add(len(request.Assets))
+	for idx, asset := range request.Assets {
+		f := func(idx int, asset *validator.Asset) func() {
+			return func() {
+				defer requestDone.Done()
+				if err := asset2.ValidateAsset(asset); err != nil {
+					errs.Add(errors.Wrapf(err, "index %d", idx))
+					return
+				}
+
+				f, err := asset2.ConvertResourceViaJSONToInterface(asset)
+				if err != nil {
+					errs.Add(errors.Wrapf(err, "index %d", idx))
+					return
+				}
+
+				violations, err := v.constraintFramework.Review(ctx, f)
+				if err != nil {
+					return
+				}
+
+				if len(violations) != 0 {
+					responseMutex.Lock()
```

Comment:
Is a Mutex the best option here?

Might it be more efficient to have workers simply push violations onto a channel which the main function reads from to return the review response?

Reply:
That's an interesting question.  After thinking about this for a few minutes, I think you may be right that a channel would be more efficient, however, it's one of those things where we don't know for sure until we benchmark it.  In any case, the channel would be more "go-like" so that sounds good to me.